theano/compile/builders.py

```diff
@@ -115,40 +73,73 @@ def __hash__(self):
         # TODO: use internal variables in hash
         return hash(type(self))
 
+    def grad(self, inputs, output_grads):
+        if self.grad_op_is_cached:
+            return self.grad_op(inputs, output_grads)
+
+        if self.grad_op is None:
+            self.grad_op = []
+
+        # we need to convert a list into a single funtor
+        if isinstance(self.grad_op, list):
+            grad_op_l = self.grad_op
+            if len(grad_op_l) > len(self.local_inputs):
+                raise ValueError(
+                    'Can override %d gradients at most, got %d' % (
+                        len(self.local_inputs), len(grad_op_l)))
+            if len(grad_op_l) < len(self.local_inputs):
+                grad_op_l += [None] * (
+                    len(self.local_inputs) - len(grad_op_l))
+            wrt = [self.local_inputs[i] for i, go in
+                   enumerate(grad_op_l) if not go]
+            # compute non-overriding downsteam gradients from upstreams grads
+            # it's normal some input may be disconnected, thus the 'ignore'
+            ups_grads_d = dict(izip(self.local_outputs, output_grads))
+            nat_dns_grads = iter(theano.gradient.grad(
+                cost=None,
+                known_grads=ups_grads_d,
+                wrt=wrt,
+                disconnected_inputs='ignore'))
+            # combine overriding gradients
+            dns_grads_l = [
+                go(self.local_inputs, output_grads) if go else next(nat_dns_grads) for go in grad_op_l]
+            grad_ofg = type(self)(
+                inputs=self.local_inputs + output_grads,
+                outputs=dns_grads_l,
+                inline=self.is_inline, on_unused_input='ignore')
```

Comment:
This will still create an `OpFromGraph` even for inline Op gradients, but on the positive side, this reduces code complexity. Otherwise we'd have to duplicate the cloning code here for the inline case. I'm not opposed to this any more.